File: 0315_Count_of_Smaller_Numbers_After_Self/try_6.py
class BIT:
    def __init__(self, n):
        self.arr = [0] * (n+1)
    
    def update(self, idx, delta=1):
        while idx < len(self.arr):
            self.arr[idx] += delta
            idx += idx & -idx
    
    def query(self, idx):
        ret = 0
        while idx > 0:
            ret += self.arr[idx]
            idx -= idx & -idx
        return ret

class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        # reorg input array
        n = len(nums)
        _min = min(nums)
        for i in range(n):
            if _min >= 0:
                nums[i] += _min
            else:
                nums[i] -= _min
        
        # get ans
        ans = []
        tree = BIT(2*(10**4))
        for num in nums[::-1]:
            ans.append(tree.query(num))
            tree.update(num+1)
        return ans[::-1]
        

File: 0053_Maximum_Subarray/try_6.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int ans {INT_MIN}, cur_sum {0};
        
        for (auto& num : nums) {
            cur_sum += num;
            ans = max(ans, cur_sum);
            
            if (cur_sum < 0) {
                cur_sum = 0;
            }
        }
        
        return ans;
    }
};
